lmfit/minimizer.py

```diff
@@ -499,6 +507,124 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
+            How many samples you would like to draw from the posterior
+            distribution?
+        burn : int, optional
+            Discard this many samples from the start of the sampling regime.
+        thin : int, optional
+            Only accept 1 in every `thin` samples.
+        ntemps : int, optional
+            If `ntemps > 1` perform a Parallel Tempering.
+
+        Returns
+        -------
+        result, chain : MinimizerResult, pandas.DataFrame
+            MinimizerResult object contains updated params, fit statistics, etc.
+            The `chain` contains the samples. Has shape (steps, ndims). Access
+            chain values for a particular parameter using chain[parname].
+        """
+        if not HAS_EMCEE:
+            raise NotImplementedError('You must have emcee to use the emcee '
+                                      'method')
+
+        result = self.prepare_fit(params=params)
+        vars   = result.init_vals
+        params = result.params
+
+        # Removing internal parameter scaling. We could possibly keep it,
+        # but I don't know how this affects the emcee sampling.
+        bounds_varying = []
+        for i, par in enumerate(params):
+            param = params[par]
+            vars[i] = param.from_internal(param.value)
+            param.from_internal = lambda val: val
+            lb, ub = param.min, param.max
+            if lb is None or lb is np.nan:
+                lb = -np.inf
+            if ub is None or ub is np.nan:
+                ub = np.inf
+            bounds_varying.append((lb, ub))
+
+        bounds_varying = np.array(bounds_varying)
+
+        self.nvarys = len(result.var_names)
+
+        def lnlike(theta):
+            # log likelihood
+            return -0.5 * self.penalty(theta)
+
+        def lnprior(theta):
+            # stay within the prior specified by the parameter bounds
+            if (np.any(theta > bounds_varying[:, 1])
+                    or np.any(theta < bounds_varying[:, 0])):
+                return -np.inf
+            return 0
+
+        def lnprob(theta):
+            lp = lnprior(theta)
+            if not np.isfinite(lp):
+                return -np.inf
+            return lp + lnlike(theta)
+
+        if ntemps > 1:
+            # jitter the starting position by scaled Gaussian noise (1% level)
+            p0 = 1 + np.random.randn(ntemps, nwalkers, self.nvarys) * 1.e-2
+            sampler = emcee.PTSampler(ntemps, nwalkers, self.nvarys, lnlike,
+                                      lnprior)
+        else:
+            p0 = 1 + np.random.randn(nwalkers, self.nvarys) * 1.e-2
+            sampler = emcee.EnsembleSampler(nwalkers, self.nvarys, lnprob)
+
+        # burn in the sampler
+        p0 *= vars
+        for output in sampler.sample(p0, iterations=burn):
+            p0 = output[0]
+        sampler.reset()
+
+        # now do a production run
+        for output in sampler.sample(p0, iterations=steps - burn, thin=thin):
```

Comment:
Using `run_mcmc` would remove the for loop.  However, it still would need to be run twice, so that the burn samples could be discarded. Neither `run_mcmc` nor `sample` have a `burn` keyword. THe burn samples could be discarded from the chain in post processing, but the chain output from the serial and parallel samplers has different dimensionalities - one is 3D the other 4D. Which makes handling messier.